pkg/controller/operators/openshift/helpers.go

```diff
@@ -0,0 +1,234 @@
+		return object.GetNamespace() == *namespace
+	})
+}
+
+func watchName(name *string) predicate.Funcs {
+	return predicate.NewPredicateFuncs(func(object client.Object) bool {
+		return object.GetName() == *name
+	})
+}
+
+func conditionsEqual(a, b *configv1.ClusterOperatorStatusCondition) bool {
+	if a == b {
+		return true
+	}
+
+	if a == nil || b == nil {
+		return false
+	}
+
+	return a.Type == b.Type && a.Status == b.Status && a.Message == b.Message && a.Reason == b.Reason
+}
+
+type skews []skew
+
+func (s skews) String() string {
+	// TODO: Use a string builder.
+	str := "The following operators block OpenShift upgrades: "
+	for i, sk := range s {
+		if i < len(s)-1 {
+			str += sk.String() + ", "
+			continue
+		}
+		str += sk.String()
+	}
+
+	return str
+}
+
+type skew struct {
+	namespace           string
+	name                string
+	maxOpenShiftVersion string
+}
+
+func (s skew) String() string {
+	return fmt.Sprintf("Operator %s in namespace %s is not compatible with OpenShift versions greater than %s", s.name, s.namespace, s.maxOpenShiftVersion)
+}
+
+func incompatibleOperators(ctx context.Context, cli client.Client) (skews, error) {
+	next, err := desiredRelease(ctx, cli)
+	if err != nil {
+		return nil, err
+	}
+	next.Minor++
+
+	csvList := &operatorsv1alpha1.ClusterServiceVersionList{}
+	if err := cli.List(ctx, csvList); err != nil {
+		return nil, err
+	}
+
+	var (
+		s    skews
+		errs []error
+	)
+	for _, csv := range csvList.Items {
+		if csv.IsCopied() {
+			continue
+		}
+
+		max, err := maxOpenShiftVersion(&csv)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if max == nil || max.GTE(*next) {
+			continue
+		}
+
+		s = append(s, skew{
+			name:                csv.GetName(),
+			namespace:           csv.GetNamespace(),
+			maxOpenShiftVersion: max.String(),
+		})
+	}
+
+	return s, utilerrors.NewAggregate(errs)
+}
+
+func desiredRelease(ctx context.Context, cli client.Client) (*semver.Version, error) {
+	cv := configv1.ClusterVersion{}
+	if err := cli.Get(ctx, client.ObjectKey{Name: "version"}, &cv); err != nil { // "version" is the name of OpenShift's ClusterVersion singleton
+		return nil, err
+	}
+
+	v := cv.Status.Desired.Version
+	if v == "" {
+		// The release version hasn't been set yet
+		return nil, nil
+	}
+
+	desired, err := semver.ParseTolerant(v)
+	if err != nil {
+		return nil, err
+	}
+
+	return &desired, nil
+}
+
+const (
+	MaxOpenShiftVersionProperty = "olm.maxOpenShiftVersion"
+)
+
+func maxOpenShiftVersion(csv *operatorsv1alpha1.ClusterServiceVersion) (*semver.Version, error) {
+	// Extract the property from the CSV's annotations if possible
+	annotation, ok := csv.GetAnnotations()[projection.PropertiesAnnotationKey]
+	if !ok {
+		return nil, nil
+	}
+
+	properties, err := projection.PropertyListFromPropertiesAnnotation(annotation)
+	if err != nil {
+		return nil, err
+	}
+
+	// Take the highest semver if there's more than one max version specified
```

Comment:
Would it be a safer choice to go with the lowest version?

Reply:
I was going at this from the perspective that folks would always make a conscious decision to add the property, but over several releases may forget to remove older ones.